h2o-core/src/main/java/hex/Model.java

```diff
@@ -1321,9 +1322,14 @@ public double lift_top_group() {
             computeMetrics,
             _output.interactionBuilder(),
             getToEigenVec(),
-            _toDelete,
+            toDelete,
             catEncoded
     );
+    for (Map.Entry<Key, String> entry: _toDelete.entrySet())
+      toDelete.putIfAbsent(entry.getKey(), entry.getValue());
+
+    _toDelete = toDelete;
```

Comment:
why not _toDelete.putAll(toDelete)?

Reply:
That fails the same way on `assert !_write_lock;`. I'm not sure but it seem to me that the potential issue comes from overwriting something that exists in the map while it is in `IcedHashMapBase::write_impl`. (I'm guessing that is the reason why `putIfAbsent()` doesn't check the `_write_lock`.)

Reply:
Problem are the parallel operations on top of one map. Where do they come from ? In my case clone helps. Note to self: See how Clone is implemented. This seems either like a:

1. Bug in ModelBuilder
2. Intended state but poor H2O locking mechanism (this structure should be guaranteed to be used lock-free, but it's not).

Reply:
If this is about concurent modifications of _toDelete than I dont think this is the right way to go because this can throw away modifications done to _toDelete in parallel. Overall I understand this is a workaround to get something fixed but I would much rather see a proper fix to this problem, even though I understand this might be very difficult.

Reply:
Thanks, that's a very good point! I am still not able to determine what's actually causing it but I was able to make a different fix that shouldn't suffer the "throwing away modifications" issue.

In `adaptTestForTrain`, I changed `toDelete.put` to `toDelete.putIfAbsent`.  That shouldn't have any negative side effect as far as I can tell. 

As far as I can tell the issue isn't about concurrent modification, it seem to be about being modified while being read in `IcedHashMapBase::write_impl`.

Docstring of `write_impl` says:
```
  // Since this is a *concurrent* hashtable, writing it whilst its being
  // updated is tricky.  If the table is NOT being updated, then all is written
  // as expected.  If the table IS being updated we only promise to write the
  // Keys that existed at the time the table write began.  If elements are
  // being deleted, they may be written anyways.  If the Values are changing, a
  // random Value is written.
```

So to me seems that this fix should be ok => if the new entry isn't written in current run of `write_impl` it should be written the next time. Since  `putIfAbsent` just adds, there shouldn't be any inconsistencies.